File: packages/foresight-core/src/crdt.ts
/**
 * CRDT (Conflict-Free Replicated Data Type) Implementations
 */
import { VectorClockType } from './types'

// =============================================================================
// Vector Clock
// =============================================================================

export class VectorClock {
  private clock: Record<string, number>

  constructor(data?: Record<string, number>) {
    this.clock = data ? { ...data } : {}
  }

  increment(nodeId: string): void {
    this.clock[nodeId] = (this.clock[nodeId] || 0) + 1
  }

  merge(other: VectorClock): void {
    const otherClock = other.toDict()
    const allNodes = new Set([
      ...Object.keys(this.clock),
      ...Object.keys(otherClock),
    ])

    for (const nodeId of allNodes) {
      this.clock[nodeId] = Math.max(
        this.clock[nodeId] || 0,
        otherClock[nodeId] || 0,
      )
    }
  }

  happensBefore(other: VectorClock): boolean {
    const otherClock = other.toDict()
    const allNodes = new Set([
      ...Object.keys(this.clock),
      ...Object.keys(otherClock),
    ])

    let atLeastOneLess = false
    for (const nodeId of allNodes) {
      const selfVal = this.clock[nodeId] || 0
      const otherVal = otherClock[nodeId] || 0

      if (selfVal > otherVal) {
        return false
      }
      if (selfVal < otherVal) {
        atLeastOneLess = true
      }
    }

    return atLeastOneLess
  }

  concurrentWith(other: VectorClock): boolean {
    return !this.happensBefore(other) && !other.happensBefore(this)
  }

  copy(): VectorClock {
    return new VectorClock(this.clock)
  }

  toDict(): Record<string, number> {
    return { ...this.clock }
  }

  static fromDict(data: Record<string, number>): VectorClock {
    return new VectorClock(data)
  }
}

// =============================================================================
// LWW-Register (Last-Writer-Wins Register)
// =============================================================================

export class LWWRegister<T> {
  private value: T | null
  private timestamp: number
  private nodeId: string
  private readonly vectorClock: VectorClock

  constructor(
    value: T | null = null,
    timestamp: number = 0,
    nodeId: string = '',
    vectorClock: VectorClock = new VectorClock(),
  ) {
    this.value = value
    this.timestamp = timestamp
    this.nodeId = nodeId
    this.vectorClock = vectorClock
  }

  set(newValue: T, nodeId: string, timestamp?: number): void {
    const ts = timestamp ?? Date.now() / 1000

    // Accept if timestamp is newer, or same timestamp and nodeId is >= current
    if (
      ts > this.timestamp ||
      (ts === this.timestamp && nodeId >= this.nodeId)
    ) {
      this.value = newValue
      this.timestamp = ts
      this.nodeId = nodeId
      this.vectorClock.increment(nodeId)
    }
  }

  merge(other: LWWRegister<T>): void {
    if (other.timestamp > this.timestamp) {
      this.value = other.value
      this.timestamp = other.timestamp
      this.nodeId = other.nodeId
      this.vectorClock.merge(other.vectorClock)
    } else if (other.timestamp === this.timestamp) {
      if (other.nodeId > this.nodeId) {
        this.value = other.value
        this.nodeId = other.nodeId
      }
      this.vectorClock.merge(other.vectorClock)
    } else {
      this.vectorClock.merge(other.vectorClock)
    }
  }

  get(): T | null {
    return this.value
  }

  toDict(): { value: T | null; timestamp: number; nodeId: string; vectorClock: Record<string, number> } {
    return {
      value: this.value,
      timestamp: this.timestamp,
      nodeId: this.nodeId,
      vectorClock: this.vectorClock.toDict(),
    }
  }

  static fromDict<T>(data: { value: T | null; timestamp?: number; nodeId?: string; vectorClock?: Record<string, number> }): LWWRegister<T> {
    return new LWWRegister<T>(
      data.value,
      data.timestamp ?? 0,
      data.nodeId ?? '',
      data.vectorClock
        ? VectorClock.fromDict(data.vectorClock)
        : new VectorClock(),
    )
  }
}

// =============================================================================
// OR-Set (Observed-Remove Set)
// =============================================================================

export class ORSet<T> {
  private readonly adds: Map<string, Set<string>> // hash -> Set of "timestamp:nodeId"
  private readonly removes: Map<string, Set<string>> // hash -> Set of "timestamp:nodeId"
  private nodeId: string = 'default'
  private vc: VectorClock = new VectorClock()

  constructor() {
    this.adds = new Map()
    this.removes = new Map()
  }

  setNodeId(nodeId: string): void {
    this.nodeId = nodeId
  }

  get vectorClock(): VectorClock {
    return this.vc
  }

  private getHash(element: T): string {
    const str = typeof element === 'string' ? element : JSON.stringify(element)
    let hash = 0
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i)
      hash = (hash << 5) - hash + char
      hash |= 0 // Convert to 32bit integer
    }
    return hash.toString(16)
  }

  /**
   * Performs garbage collection by removing tags that are in both adds and removes.
   * If all adds are removed, the element hash entry is deleted.
   */
  cleanup(): void {
    for (const [hash, adds] of this.adds.entries()) {
      const removes = this.removes.get(hash)
      if (removes) {
        for (const tag of adds) {
          if (removes.has(tag)) {
            adds.delete(tag)
            removes.delete(tag)
          }
        }
        if (removes.size === 0) {
          this.removes.delete(hash)
        }
      }
      if (adds.size === 0) {
        this.adds.delete(hash)
      }
    }
  }

  add(element: T): void {
    const ts = Date.now() / 1000
    const elementHash = this.getHash(element)
    const tag = `${ts}:${this.nodeId}`

    if (!this.adds.has(elementHash)) {
      this.adds.set(elementHash, new Set())
    }
    this.adds.get(elementHash)!.add(tag)

    this.vc.increment(this.nodeId)
  }

  remove(element: T): void {
    const elementHash = this.getHash(element)

    // Mark all current adds as removed
    if (this.adds.has(elementHash)) {
      if (!this.removes.has(elementHash)) {
        this.removes.set(elementHash, new Set())
      }
      for (const tag of this.adds.get(elementHash)!) {
        this.removes.get(elementHash)!.add(tag)
      }
    }

    this.vc.increment(this.nodeId)
  }

  contains(element: T): boolean {
    const elementHash = this.getHash(element)
    return this.containsHash(elementHash)
  }

  private containsHash(elementHash: string): boolean {
    if (!this.adds.has(elementHash)) {
      return false
    }

    const adds = this.adds.get(elementHash)!
    const removes = this.removes.get(elementHash) ?? new Set()

    // Element is in set if there's at least one add not matched by remove
    for (const tag of adds) {
      if (!removes.has(tag)) {
        return true
      }
    }

    return false
  }

  /**
   * Returns elements as hashes.
   * Note: In a full implementation, you would store the values alongside the hashes.
   */
  getElements(): string[] {
    const elements: string[] = []
    for (const hash of this.adds.keys()) {
      if (this.containsHash(hash)) {
        elements.push(hash)
      }
    }
    return elements
  }

  merge(other: ORSet<T>): void {
    const otherDict = other.toDict()

    // Merge adds
    for (const [hash, tags] of Object.entries(otherDict.adds)) {
      if (!this.adds.has(hash)) {
        this.adds.set(hash, new Set())
      }
      for (const tag of tags) {
        this.adds.get(hash)!.add(tag)
      }
    }

    // Merge removes
    for (const [hash, tags] of Object.entries(otherDict.removes)) {
      if (!this.removes.has(hash)) {
        this.removes.set(hash, new Set())
      }
      for (const tag of tags) {
        this.removes.get(hash)!.add(tag)
      }
    }

    this.vc.merge(other.vectorClock)
  }

  toDict(): { adds: Record<string, string[]>; removes: Record<string, string[]>; vectorClock: Record<string, number>; nodeId: string } {
    const addsObj: Record<string, string[]> = {}
    for (const [hash, tags] of this.adds.entries()) {
      addsObj[hash] = Array.from(tags)
    }

    const removesObj: Record<string, string[]> = {}
    for (const [hash, tags] of this.removes.entries()) {
      removesObj[hash] = Array.from(tags)
    }

    return {
      adds: addsObj,
      removes: removesObj,
      vectorClock: this.vc.toDict(),
      nodeId: this.nodeId,
    }
  }

  static fromDict<T>(data: { nodeId?: string; vectorClock?: Record<string, number>; adds?: Record<string, string[]>; removes?: Record<string, string[]> }): ORSet<T> {
    const orset = new ORSet<T>()
    orset.nodeId = data.nodeId ?? 'default'
    if (data.vectorClock) {
      orset.vc = VectorClock.fromDict(data.vectorClock)
    }

    if (data.adds) {
      for (const [hash, tags] of Object.entries(data.adds)) {
        orset.adds.set(hash, new Set(tags))
      }
    }

    if (data.removes) {
      for (const [hash, tags] of Object.entries(data.removes)) {
        orset.removes.set(hash, new Set(tags))
      }
    }

    return orset
  }
}

// =============================================================================
// LWW-Map (Last-Writer-Wins Map)
// =============================================================================

export class LWWMap<T> {
  private readonly entries: Map<string, LWWRegister<T | null>>
  private nodeId: string = 'default'
  private vc: VectorClock = new VectorClock()

  constructor() {
    this.entries = new Map()
  }

  setNodeId(nodeId: string): void {
    this.nodeId = nodeId
  }

  get vectorClock(): VectorClock {
    return this.vc
  }

  set(key: string, value: T): void {
    if (!this.entries.has(key)) {
      this.entries.set(key, new LWWRegister<T | null>())
    }
    this.entries.get(key)!.set(value, this.nodeId)
    this.vc.increment(this.nodeId)
  }

  get(key: string): T | null {
    if (!this.entries.has(key)) {
      return null
    }
    return this.entries.get(key)!.get()
  }

  /**
   * Removes null entries (tombstones).
   * Note: In a distributed system, this should be used with caution as it might
   * cause deleted items to reappear if merged with a node that hasn't seen the deletion.
   */
  cleanup(): void {
    for (const [key, reg] of this.entries.entries()) {
      if (reg.get() === null) {
        this.entries.delete(key)
      }
    }
  }

  delete(key: string): void {
    if (this.entries.has(key)) {
      // Set to null with new timestamp (tombstone)
      this.entries.get(key)!.set(null, this.nodeId)
      this.vc.increment(this.nodeId)
    }
  }

  keys(): string[] {
    return Array.from(this.entries.keys())
  }

  merge(other: LWWMap<T>): void {
    const otherDict = other.toDict()

    for (const [key, regData] of Object.entries(otherDict.entries)) {
      const otherRegister = LWWRegister.fromDict<T | null>(regData)
      if (!this.entries.has(key)) {
        this.entries.set(key, otherRegister)
      } else {
        this.entries.get(key)!.merge(otherRegister)
      }
    }
    this.vc.merge(other.vectorClock)
  }

  toDict(): { entries: Record<string, { value: T | null; timestamp: number; nodeId: string; vectorClock: Record<string, number> }>; vectorClock: Record<string, number>; nodeId: string } {
    const entriesObj: Record<string, { value: T | null; timestamp: number; nodeId: string; vectorClock: Record<string, number> }> = {}
    for (const [key, reg] of this.entries.entries()) {
      entriesObj[key] = reg.toDict()
    }

    return {
      entries: entriesObj,
      vectorClock: this.vc.toDict(),
      nodeId: this.nodeId,
    }
  }

  static fromDict<T>(data: { nodeId?: string; vectorClock?: Record<string, number>; entries?: Record<string, { value: T | null; timestamp?: number; nodeId?: string; vectorClock?: Record<string, number> }> }): LWWMap<T> {
    const lwwMap = new LWWMap<T>()
    lwwMap.nodeId = data.nodeId ?? 'default'
    if (data.vectorClock) {
      lwwMap.vc = VectorClock.fromDict(data.vectorClock)
    }

    if (data.entries) {
      for (const [key, regData] of Object.entries(data.entries)) {
        lwwMap.entries.set(key, LWWRegister.fromDict<T | null>(regData))
      }
    }

    return lwwMap
  }
}

File: foresight_mcp/crdt.py
"""
CRDT (Conflict-Free Replicated Data Type) Implementations

Provides conflict-free data types for offline-first synchronization:
- LWW-Register: Last-Writer-Wins Register for scalar values
- OR-Set: Observed-Remove Set for collections
- Vector Clock: Causal ordering of events
"""

from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass, field
from typing import Any, TypeVar

T = TypeVar("T")


# =============================================================================
# Vector Clock
# =============================================================================


@dataclass
class VectorClock:
    """
    Vector clock for causal ordering.

    Tracks logical timestamps across distributed nodes.
    Each node increments its own counter when performing operations.

    Attributes:
        clock: Dict mapping node_id to counter
    """

    clock: dict[str, int] = field(default_factory=dict)

    def increment(self, node_id: str) -> None:
        """Increment clock for a node."""
        self.clock[node_id] = self.clock.get(node_id, 0) + 1

    def merge(self, other: VectorClock) -> None:
        """Merge with another vector clock (take max of each component)."""
        all_nodes = set(self.clock.keys()) | set(other.clock.keys())
        for node_id in all_nodes:
            self.clock[node_id] = max(self.clock.get(node_id, 0), other.clock.get(node_id, 0))

    def happens_before(self, other: VectorClock) -> bool:
        """
        Check if this clock happens-before another.

        A happens-before B if:
        - All components of A <= corresponding components of B
        - At least one component of A < corresponding component of B
        """
        all_nodes = set(self.clock.keys()) | set(other.clock.keys())

        at_least_one_less = False
        for node_id in all_nodes:
            self_val = self.clock.get(node_id, 0)
            other_val = other.clock.get(node_id, 0)

            if self_val > other_val:
                return False
            if self_val < other_val:
                at_least_one_less = True

        return at_least_one_less

    def concurrent_with(self, other: VectorClock) -> bool:
        """Check if this clock is concurrent with another (neither happens-before)."""
        return not self.happens_before(other) and not other.happens_before(self)

    def copy(self) -> VectorClock:
        """Create a copy of this vector clock."""
        new_clock = VectorClock()
        new_clock.clock = self.clock.copy()
        return new_clock

    def to_dict(self) -> dict[str, int]:
        """Convert to dictionary."""
        return self.clock.copy()

    @classmethod
    def from_dict(cls, data: dict[str, int]) -> VectorClock:
        """Create from dictionary."""
        clock = VectorClock()
        clock.clock = data
        return clock


# =============================================================================
# LWW-Register (Last-Writer-Wins Register)
# =============================================================================


@dataclass
class LWWRegister[T]:
    """
    Last-Writer-Wins Register CRDT.

    Resolves conflicts by choosing the value with the latest timestamp.
    Ties are broken by node_id (lexicographically higher wins).

    Attributes:
        value: Current value
        timestamp: Logical timestamp (Unix timestamp)
        node_id: ID of the node that wrote the value
        vector_clock: Causal ordering
    """

    value: T | None = None
    timestamp: float = 0.0
    node_id: str = ""
    vector_clock: VectorClock = field(default_factory=VectorClock)

    def set(self, new_value: T, node_id: str, timestamp: float | None = None) -> None:
        """
        Set a new value.

        Args:
            new_value: Value to set
            node_id: ID of the node performing the operation
            timestamp: Optional timestamp (defaults to current time)
        """
        ts = timestamp or time.time()

        # Always accept if timestamp is newer, or same timestamp and node_id is higher
        if ts > self.timestamp or (ts == self.timestamp and node_id > self.node_id):
            self.value = new_value
            self.timestamp = ts
            self.node_id = node_id
            self.vector_clock.increment(node_id)

    def merge(self, other: LWWRegister[T]) -> None:
        """Merge with another LWW-Register (last writer wins)."""
        if other.timestamp > self.timestamp:
            self.value = other.value
            self.timestamp = other.timestamp
            self.node_id = other.node_id
            self.vector_clock.merge(other.vector_clock)
        elif other.timestamp == self.timestamp:
            if other.node_id > self.node_id:
                self.value = other.value
                self.node_id = other.node_id
            self.vector_clock.merge(other.vector_clock)
        else:
            self.vector_clock.merge(other.vector_clock)

    def get(self) -> T | None:
        """Get the current value."""
        return self.value

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "value": self.value,
            "timestamp": self.timestamp,
            "node_id": self.node_id,
            "vector_clock": self.vector_clock.to_dict(),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> LWWRegister:
        """Create from dictionary."""
        register = cls()
        register.value = data.get("value")
        register.timestamp = data.get("timestamp", 0.0)
        register.node_id = data.get("node_id", "")
        if "vector_clock" in data:
            register.vector_clock = VectorClock.from_dict(data["vector_clock"])
        return register


# =============================================================================
# OR-Set (Observed-Remove Set)
# =============================================================================


@dataclass
class ORSet[T]:
    """
    Observed-Remove Set CRDT.

    Elements can be added and removed concurrently without conflicts.
    An element is in the set if it has been added and not subsequently removed.

    Attributes:
        _adds: Dict mapping element hash to set of (timestamp, node_id) pairs
        _removes: Dict mapping element hash to set of (timestamp, node_id) pairs
        _values: Dict mapping element hash to the most recent element value
        _cache: Cache of current set contents
    """

    _adds: dict[str, set[tuple]] = field(default_factory=dict)
    _removes: dict[str, set[tuple]] = field(default_factory=dict)
    _values: dict[str, T] = field(default_factory=dict)
    _cache: set[T] | None = None
    _node_id: str = "default"
    vector_clock: VectorClock = field(default_factory=VectorClock)

    def __post_init__(self):
        if self._adds is None:
            self._adds = {}
        if self._removes is None:
            self._removes = {}
        if self._values is None:
            self._values = {}

    def set_node_id(self, node_id: str) -> None:
        """Set the node ID for this replica."""
        self._node_id = node_id

    def _get_hash(self, element: T) -> str:
        """Get hash of an element."""
        return hashlib.sha256(str(element).encode()).hexdigest()[:16]

    def add(self, element: T) -> None:
        """Add an element to the set."""
        ts = time.time()
        element_hash = self._get_hash(element)
        tag = (ts, self._node_id)

        if element_hash not in self._adds:
            self._adds[element_hash] = set()
        self._adds[element_hash].add(tag)
        self._values[element_hash] = element

        self.vector_clock.increment(self._node_id)
        self._cache = None

    def remove(self, element: T) -> None:
        """Remove an element from the set."""
        element_hash = self._get_hash(element)

        # Mark all current adds as removed
        if element_hash in self._adds:
            if element_hash not in self._removes:
                self._removes[element_hash] = set()
            # Copy adds to removes
            for add_tag in self._adds[element_hash]:
                self._removes[element_hash].add(add_tag)

        self.vector_clock.increment(self._node_id)
        self._cache = None

    def contains(self, element: T) -> bool:
        """Check if element is in the set."""
        element_hash = self._get_hash(element)

        if element_hash not in self._adds:
            return False

        # Element is in set if there's at least one add not matched by remove
        adds = self._adds.get(element_hash, set())
        removes = self._removes.get(element_hash, set())

        return len(adds - removes) > 0

    def get_elements(self) -> set[T]:
        """Get all elements in the set."""
        if self._cache is not None:
            return self._cache

        self._cache = {
            self._values[element_hash]
            for element_hash, adds in self._adds.items()
            if element_hash in self._values and adds - self._removes.get(element_hash, set())
        }
        return self._cache

    def merge(self, other: ORSet[T]) -> None:
        """Merge with another OR-Set."""
        # Merge adds
        for element_hash, tags in other._adds.items():
            if element_hash not in self._adds:
                self._adds[element_hash] = set()
            self._adds[element_hash].update(tags)
            if element_hash in other._values:
                self._values[element_hash] = other._values[element_hash]

        # Merge removes
        for element_hash, tags in other._removes.items():
            if element_hash not in self._removes:
                self._removes[element_hash] = set()
            self._removes[element_hash].update(tags)

        self.vector_clock.merge(other.vector_clock)
        self._cache = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "adds": {k: list(v) for k, v in self._adds.items()},
            "removes": {k: list(v) for k, v in self._removes.items()},
            "values": self._values.copy(),
            "vector_clock": self.vector_clock.to_dict(),
            "node_id": self._node_id,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ORSet:
        """Create from dictionary."""
        orset = cls()
        orset._adds = {k: set(v) for k, v in data.get("adds", {}).items()}
        orset._removes = {k: set(v) for k, v in data.get("removes", {}).items()}
        orset._values = data.get("values", {}).copy()
        orset._node_id = data.get("node_id", "default")
        if "vector_clock" in data:
            orset.vector_clock = VectorClock.from_dict(data["vector_clock"])
        return orset


# =============================================================================
# LWW-Map (Last-Writer-Wins Map)
# =============================================================================


@dataclass
class LWWMap[T]:
    """
    Last-Writer-Wins Map CRDT.

    A map where each key is an LWW-Register.
    Concurrent writes to the same key are resolved by LWW.

    Attributes:
        _entries: Dict mapping keys to LWW-Registers
    """

    _entries: dict[str, LWWRegister] = field(default_factory=dict)
    _node_id: str = "default"
    vector_clock: VectorClock = field(default_factory=VectorClock)

    def set_node_id(self, node_id: str) -> None:
        """Set the node ID for this replica."""
        self._node_id = node_id

    def set(self, key: str, value: T) -> None:
        """Set a key-value pair."""
        if key not in self._entries:
            self._entries[key] = LWWRegister()
        self._entries[key].set(value, self._node_id)
        self.vector_clock.increment(self._node_id)

    def get(self, key: str) -> T | None:
        """Get value for a key."""
        if key not in self._entries:
            return None
        return self._entries[key].get()

    def delete(self, key: str) -> None:
        """Delete a key (tombstone)."""
        if key in self._entries:
            # Set to None with new timestamp
            self._entries[key].set(None, self._node_id)
            self.vector_clock.increment(self._node_id)

    def keys(self) -> list[str]:
        """Get all keys."""
        return list(self._entries.keys())

    def merge(self, other: LWWMap[T]) -> None:
        """Merge with another LWW-Map."""
        for key, other_register in other._entries.items():
            if key not in self._entries:
                self._entries[key] = LWWRegister.from_dict(other_register.to_dict())
            else:
                self._entries[key].merge(other_register)
        self.vector_clock.merge(other.vector_clock)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "entries": {k: v.to_dict() for k, v in self._entries.items()},
            "vector_clock": self.vector_clock.to_dict(),
            "node_id": self._node_id,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> LWWMap:
        """Create from dictionary."""
        lww_map = cls()
        entries = data.get("entries", {})
        lww_map._entries = {k: LWWRegister.from_dict(v) for k, v in entries.items()}
        lww_map._node_id = data.get("node_id", "default")
        if "vector_clock" in data:
            lww_map.vector_clock = VectorClock.from_dict(data["vector_clock"])
        return lww_map
